### ISSUE TITLE
uvicorn can not start FastAPI with example settings

### ISSUE THREAD

**Issue by cgi1** [tiangolo/fastapi#1495]
```
import uvicorn
from fastapi import FastAPI
app = FastAPI(title='MADS API')
uvicorn.run(app, host='0.0.0.0', port=8127, workers=2)
WARNING:  You must pass the application as an import string to enable 'reload' or 'workers'.
```

than it quits with SystemError 1

[Corresponding ticket in uvicorn](https://github.com/encode/uvicorn/issues/687)

Versions:
```
$ pip3 list |grep "uvicorn\|fastapi"
fastapi                          0.55.1       
uvicorn                          0.11.5 
```

**Comment by jorgerpo** [comment]
Why do you say it is a bug?
It is your first try at FastAPI?

**Comment by cgi1** [comment]
No, i upgraded from Ubuntu 18.04 to Ubuntu 20.04. It was running fine on Ubuntu 18.04. productive a long time like this.

**Comment by phy25** [comment]
> You must pass the application as an import string to enable 'reload' or 'workers'.

Is the program stopping here? Looks like it's only a warning but it's very straightforward about what's going on.

**Comment by cgi1** [comment]
@phy25 Yes it stops here. I was searching through a bit, but only found [this issue report](https://github.com/encode/uvicorn/issues/484) on uvicorn under Windows.. But I think they are not really related.

**Comment by cgi1** [comment]
When I try it as supposed in the related issues, it gives me errors as well:

```
uvicorn.run("app", host='0.0.0.0', port=8127, workers=2)
ERROR:    Error loading ASGI app. Import string "app" must be in format "<module>:<attribute>".
```

and 
```
uvicorn.run("main:app", host='0.0.0.0', port=8127, workers=2)
ERROR:    Error loading ASGI app. Could not import module "main".
```

So uvicorn does not found the `FastAPI` application.

**Comment by phy25** [comment]
This is intended and not related to fastapi:

https://github.com/encode/uvicorn/blob/9b92925a352b9743c5cfcef4a65e74a81a1bad4f/uvicorn/main.py#L343

You need to run it with command line `uvicorn` if you want to run with multiple workers. Please read uvicorn documentation.

**Comment by tiangolo** [comment]
Thanks for the help here @phy25 ! :clap: :bow: 

If that solves the original problem, then you can close this issue @cgi1  :heavy_check_mark:

**Comment by cgi1** [comment]
The answer I was searching for [is here](https://github.com/encode/uvicorn/issues/687#issuecomment-637065723): You can actually start it the way I wanted it, but make sure to name it correctly:

```
uvicorn.run("main:app", host='0.0.0.0', port=8127, workers=2)
ERROR:    Error loading ASGI app. Could not import module "main".
```

comes because the **file** is not named **main.py**! If you have it as `my_fastapi_server.py`, you run it as follows:

```
uvicorn.run("my_fastapi_server:app", host='0.0.0.0', port=8127, workers=2)
```

**Comment by mrtolkien** [comment]
Adding another snippet as the ones posted here were pretty cryptic for me.

With the following files:
```
app\main.py    # Containing an app object
debug_server.py
```

I needed my `debug_server.py` to be:
```python
import uvicorn

# Importing app here makes the syntax easier
from app.main import app

if __name__ == "__main__":
    uvicorn.run("debug_server:app", host="0.0.0.0", port=80, reload=True)
```

That way the debug auto-reload server did work properly and there’s nothing related to it in the `/app` folder.

**Comment by chandresholaniya** [comment]
I got following error

$ uvicorn main:app --reload
←[32mINFO←[0m:     Uvicorn running on ←[1mhttp://127.0.0.1:8000←[0m (Press CTRL+C to quit)
←[32mINFO←[0m:     Started reloader process [←[36m←[1m22984←[0m] using ←[36m←[1mwatchgod←[0m
←[31mERROR←[0m:    Error loading ASGI app. Could not import module "main".
←[32mINFO←[0m:     Stopping reloader process [←[36m←[1m22984←[0m]

I am using uvicorn [0.13.4] on Windows 10

Kindly suggest steps to resolve it.

**Comment by waketzheng** [comment]
> The answer I was searching for [is here](https://github.com/encode/uvicorn/issues/687#issuecomment-637065723): You can actually start it the way I wanted it, but make sure to name it correctly:
> 
> ```
> uvicorn.run("main:app", host='0.0.0.0', port=8127, workers=2)
> ERROR:    Error loading ASGI app. Could not import module "main".
> ```
> 
> raises Error because the **file** is not named **main.py**! If you have it as `my_fastapi_server.py`, you run it as follows:
> 
> ```
> uvicorn.run("my_fastapi_server:app", host='0.0.0.0', port=8127, workers=2)
> ```

You can use pathlib to avoid this problem:
```py
from pathlib imprt Path

uvicorn.run(f"{Path(__file__).stem}:app", host='0.0.0.0', port=8127, workers=2)
```

**Comment by timhughes** [comment]
Try this. 

```
if __name__ == "__main__":
    uvicorn.run("__main__:app", host="0.0.0.0", port=8000, reload=True)
```

**Comment by rafaelcascalho** [comment]
What also worked for me was this:

inside the `api.py` file
```py
if __name__ == "__main__":
    run(app="api:app", reload=True, port=8080, host="0.0.0.0")
```

**Comment by timhughes** [comment]
> What also worked for me was this:
> 
> inside the `api.py` file
> ```py
> if __name__ == "__main__":
>     run(app="api:app", reload=True, port=8080, host="0.0.0.0")
> ```

This depends on your file being named `api.py`. Using `__main__.app` means you can rename your `app.py` file to anything and it will still work.

Further reading https://docs.python.org/3/library/__main__.html

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue @cgi1 ! 🍰

And thanks everyone else for the discussions here. ☕
